README.md
![Intapago](instapago-banesco.png)

Ejemplo de uso de la API de Instapago con php
---
## Requerimientos:

* Registro en http://instapago.com/
* Api key
    * Ir a https://banesco.instapago.com/Account/API
    * Generar la llave
    * Solicitar el envío de la llave publica por correo

> La llave pública es enviada al hacer el registro

#### Punto de Entrada del API
Todas las interacciones con Instapago se realizan a través de la siguiente dirección:

```
https://api.instapago.com
```

#### Crear Pago
Este método consta del envío de los datos de un pago con tarjeta de crédito a Instapago
para su autorización:

```
https://api.instapago.com/payment
Método: POST
```

#### FormData:

* KeyId (Requerido): Llave generada desde Instapago.
* PublicKeyId (Requerido): Llave compartida Enviada por correo al crear una cuenta
en instapago.
* Amount (Requerido): Monto a Debitar, utilizando punto “.” Como separador decimal.
Por ejemplo: 200.00.
* Description (Requerido): Cadena de caracteres con la descripción de la operación.
* CardHolder (Requerido): Nombre del Tarjeta habiente.
* CardHolderID (Requerido): Cédula o RIF del Tarjeta habiente.
* CardNumber (Requerido): Numero de la tarjeta de crédito, sin espacios ni
separadores.
* CVC (Requerido): Código secreto de la Tarjeta de crédito.
* ExpirationDate (Requerido): Fecha de expiración de la tarjeta en el formato mostrado
en la misma MM/YYYY. Por Ejemplo: 10/2014.
* StatusId (Requerido): Estatus en el que se creará la transacción.
    * "1": Retener (pre-autorización).
    * "2": Pagar (autorización).
* IP (Requerido): Dirección IP del cliente.
* OrderNumber (Opcional): Numero de orden del pago según el comercio.
* Address (Opcional): Dirección asociada a la tarjeta, Utilizada por algunos bancos
para mayor seguridad.
* City (Opcional): Ciudad asociada a la tarjeta, Utilizada por algunos bancos para
mayor seguridad.
* ZipCode (Opcional): Código Postal asociada a la tarjeta, Utilizada por algunos
bancos para mayor seguridad.
* State (Opcional): Estado o provincia asociada a la tarjeta, Utilizada por algunos
bancos para mayor seguridad.

Respuesta:

* Tipo: ```application/json```.
* Valor:

```
{
    "success": Indica si fue procesado el pago (true o false),
    "message": Descripción de la respuesta (Máx. 200 caracteres),
    "id": Código del pago (Máx. 8 caracteres),
    "code": Código del respuesta del pago (Máx. 3 caracteres),
    "reference": Código de referencia bancaria (Máx. 6 caracteres),
    "voucher": HTML que contiene el recibo del pago,
    "ordernumber": Número de orden indicado por el comercio
}
```

#### Completar pago
Este método funciona para procesar un bloqueo o pre-autorización, para así procesarla y
hacer el cobro respectivo.

```
https://api.instapago.com/complete
Método: POST
```

#### FormData:
* KeyId (Requerido): Llave generada desde Instapago.
* PublicKeyId (Requerido): Llave compartida Enviada por correo al crear una cuenta
en Instapago.
* Id (Requerido): Identificador único del pago.
* Amount (Requerido): Monto por el cual se desea procesar el pago final.

Respuesta
* Tipo: ```application/json```
* Valor:

```
{
    "success": Indica si fue procesado el pago (true o false),
    "message": Descripción de la respuesta (Máx. 200 caracteres),
    "id": Código del pago (Máx. 8 caracteres),
    "code": Código del respuesta del pago (Máx. 3 caracteres),
    "reference": Código de referencia bancaria (Máx. 6 caracteres),
    "voucher": HTML que contiene el recibo del pago,
    "ordernumber": Número de orden indicado por el comercio
}
```
### Ejemplo de voucher

![voucher](voucher.png)

#### Anular Pago
Este método funciona para procesar una anulación de un pago, ya sea un pago o un
bloqueo.

```
https://api.instapago.com/payment
Método: DELETE
```

#### FormData:
* KeyId (Requerido): Llave generada desde Instapago.
* PublicKeyId (Requerido): Llave compartida Enviada por correo al crear una cuenta
en Instapago.
* Id (Requerido): Identificador único del pago.

Respuesta
* Tipo: ```application/json```
* Valor:

```
{
    "success": Indica si fue procesado el pago (true o false),
    "message": Descripción de la respuesta (Máx. 200 caracteres),
    "id": Código del pago (Máx. 8 caracteres),
    "code": Código del respuesta del pago (Máx. 3 caracteres),
    "reference": Código de referencia bancaria (Máx. 6 caracteres),
    "ordernumber": Número de orden indicado por el comercio
}
```

#### Consulta de Pago

Este método funciona para consultar un pago generado anteriormente.

```
https://api.instapago.com/payment
Método: GET
```

Parámetros:
* KeyId (Requerido): Llave generada desde Instapago.
* PublicKeyId (Requerido): Llave compartida Enviada por correo al crear una cuenta en Instapago.
* Id (Requerido): Identificador único del pago.

Respuesta

* Tipo: ```application/json```
* Valor:

```
{
    "success": Indica si fue procesado el pago (true o false),
    "message": Descripción de la respuesta (Máx. 200 caracteres),
    "id": Código del pago (Máx. 8 caracteres),
    "code": Código del respuesta del pago (Máx. 3 caracteres),
    "reference": Código de referencia bancaria (Máx. 6 caracteres),
    "voucher": HTML que contiene el recibo del pago,
    "ordernumber": Número de orden indicado por el comercio
}
```

## Códigos de respuesta

Para todas las transacciones realizadas bajo el API, los códigos HTTP de respuestas corresponden a los siguientes estados:

* ```201```: Pago procesado con éxito.
* ```400```: Error al validar los datos enviados (Adicionalmente se devuelve una cadena de
caracteres con la descripción del error).
* ```401```: Error de autenticación, ha ocurrido un error con las llaves utilizadas.
* ```403```: Pago Rechazado por el banco.
* ```500```: Ha Ocurrido un error interno dentro del servidor.
* ```503```: Ha Ocurrido un error al procesar los parámetros de entrada. Revise los datos
enviados y vuelva a intentarlo.

> **Importante**: Si recibe un código de respuesta diferente a los antes descritos deben ser tomados como errores de protocolo HTTP.

## Tarjetas de prueba
Para realizar las pruebas, se provee de los siguientes datos para comprobar la integración:

* Tarjetas aprobadas:
Pueden indicar cualquier valor para Cédula o RIF, Fecha de Vencimiento y CVC:

    * Visa: ```4111111111111111```
    * American Express: ```378282246310005```
    * MasterCard: ```5105105105105100```
    * Sambil: ```8244001100110011``
    * Rattan: ```8244021100110011```
    * Locatel: ```8244041100110011```

* Tarjeta rechazada:

Cualquier número de tarjeta de crédito Visa, MasterCard, American Express, Sambil, Rattan y Locatel válidas serán "Rechazadas". Esto no genera ninguna transacción con el Banco.
￼￼
### Ejemplos de respuesta:

Tarjeta Vencida:

```json
{
  "success":false,
  "message":"Tarjeta de Crédito Expirada",
  "id":null,
  "code":"400",
  "reference":null,
  "voucher":null,
  "ordernumber":null,
  "sequence":null,
  "approval":null,
  "lote":null
}
```

Problemas con las credenciales (api key):

```json
{
  "success": false,
  "message": "Error de autenticación. Revise sus credenciales.",
  "id": null,
  "code": "401",
  "reference": null,
  "voucher": null,
  "ordernumber": null,
  "sequence": null,
  "approval": null,
  "lote": null
}
```
> De tener este tipo de problema en especifico, debe cambiar el estado de `Key ID` a ativa en el apartado de "Configuración del API" de persistir el inconveniente debe comunicarse con el personal de instapago a los numeros  +58 (212) 963.99.22 / +58 (212) 963.01.22

## Herramientas de prueba

![Postman](postman.png)

>Postman: Supercharge your API workflow. Build, test, and document your APIs faster. More than a million developers already do.

https://www.getpostman.com/


CHANGELOG
* [31m88d192c[m - [32mabr4xas[m - [33m (HEAD -> master, origin/master, origin/HEAD)[m agregado documentacion de ejemplo [32m(89 seconds ago)[m
* [31m5af83f6[m - [32mabr4xas[m - [33m[m agregado documentacion de ejemplo [32m(2 minutes ago)[m
* [31m928798b[m - [32mabr4xas[m - [33m[m actualizando repo [32m(2 minutes ago)[m
* [31m3f112fc[m - [32mabr4xas[m - [33m[m actualizado [32m(3 minutes ago)[m
* [31mfeaa51d[m - [32mabr4xas[m - [33m[m eliminado [32m(4 minutes ago)[m
* [31m1220ae4[m - [32mabr4xas[m - [33m[m agregado documentacion de ejemplo [32m(16 minutes ago)[m
* [31m3edd7d8[m - [32mabr4xas[m - [33m[m agregado documentacion de ejemplo [32m(19 minutes ago)[m
* [31mfdd6c98[m - [32mabr4xas[m - [33m[m agregado documentacion de ejemplo [32m(20 minutes ago)[m
* [31m2b03c61[m - [32mabr4xas[m - [33m[m Agregado formulario de pruebas, nada fancy [32m(47 minutes ago)[m
* [31mfdac017[m - [32mabr4xas[m - [33m[m actualizando repo [32m(2 weeks ago)[m
* [31m1ef9e98[m - [32mabr4xas[m - [33m[m se agrega el logo de instapago y banesco [32m(2 weeks ago)[m
* [31m43e42d6[m - [32mabr4xas[m - [33m[m Actualizando Repo, se agrega voucher [32m(2 weeks ago)[m
* [31m6652e9b[m - [32mángel[m - [33m[m Update .gitignore [32m(9 weeks ago)[m
* [31m79a5cc7[m - [32mángel[m - [33m[m Update .gitignore [32m(9 weeks ago)[m
* [31m180de21[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m1c9b0d1[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31me569c0a[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m91e0297[m - [32mabr4xas[m - [33m[m typo [32m(3 months ago)[m
* [31m62a3c9c[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m963888d[m - [32mabr4xas[m - [33m[m typo [32m(3 months ago)[m
* [31mfa1a33c[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m499660d[m - [32mabr4xas[m - [33m[m typo [32m(3 months ago)[m
* [31m362f746[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m682883c[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31m44ea36f[m - [32mabr4xas[m - [33m[m Actualizando repo [32m(3 months ago)[m
* [31mc66aea2[m - [32mángel[m - [33m[m Update README.md [32m(3 months ago)[m
* [31m22e9649[m - [32mángel[m - [33m[m Create sample.php [32m(3 months ago)[m
* [31mf8a0d2d[m - [32mángel[m - [33m[m Initial commit [32m(3 months ago)[m

recursos/sample.php
<?php

/*
 * Filename: sample.php
 * Description: Ejemplo para consumo de la API de instapago
 * @author Personal de Instapago
 * @version 1.5.4
 */

$CardHolder     = trim($_POST["cardHolder"]);
$CardHolderId   = trim($_POST["cardHolderId"]);
$CardNumber     = trim($_POST["cardNumber"]);
$CVC            = trim($_POST["cardCVC"]);
$ExpirationDate = trim($_POST["cardExpiry"]);

// https://www.chriswiegman.com/2014/05/getting-correct-ip-address-php/
function get_ip() {
    if ( function_exists( 'apache_request_headers' ) ) {
        $headers = apache_request_headers();
    } else {
        $headers = $_SERVER;
    }
    if (array_key_exists( 'X-Forwarded-For', $headers ) && filter_var( $headers['X-Forwarded-For'], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 ) ) {
        $the_ip = $headers['X-Forwarded-For'];
    }elseif (array_key_exists( 'HTTP_X_FORWARDED_FOR', $headers ) && filter_var( $headers['HTTP_X_FORWARDED_FOR'], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 )) {
        $the_ip = $headers['HTTP_X_FORWARDED_FOR'];
    }else {
        $the_ip = filter_var( $_SERVER['REMOTE_ADDR'], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 );
    }
    return $the_ip;
}
$url = 'https://api.instapago.com/payment';
$fields = array(
    "KeyID"             => "", //required
    "PublicKeyId"       => "", //required
    "Amount"            => "200", //required
    "Description"       => "Un cobro de prueba", //required
    "CardHolder"        => $CardHolder, //required
    "CardHolderId"      => $CardHolderId, //required
    "CardNumber"        => $CardNumber, //required
    "CVC"               => $CVC, //required
    "ExpirationDate"    => $ExpirationDate, //required
    "StatusId"          => "2", //required
    "IP"                => get_ip() //required
);
$ch = curl_init();
curl_setopt($ch, CURLOPT_URL,$url );
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_POSTFIELDS,http_build_query($fields));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$server_output = curl_exec ($ch);
curl_close ($ch);
$obj = json_decode($server_output);
$code = $obj->code;
$msgerror = '';
switch($code){
    case 400:
        $msgerror= 'Error al validar los datos enviados..';
        break;
    case 401:
        $msgerror= 'Error de autenticación, ha ocurrido un error con las llaves utilizadas..';
        break;
    case 403:
        $msgerror= 'Pago Rechazado por el banco..';
        break;
    case 500:
        $msgerror= 'Ha Ocurrido un error interno dentro del servidor..';
        break;
    case 503:
        $msgerror= 'Ha Ocurrido un error al procesar los parámetros de entrada. Revise los datos enviados y vuelva a intentarlo..';
        break;
    case 201:
        $msg_banco  = $obj->message;
        $voucher  = $obj->voucher;
        $voucher = html_entity_decode($voucher);
        $id_pago  = $obj->id;
        $reference  = $obj->reference;
        break;
    default:
        $msgerror='Error inesperado, Imposible determinar. Contacte al Administrador';
        break;
}
if(!empty($msgerror)){
    echo'
        <div class="alert alert-danger" role="alert"><p>'.$msgerror.'</p></div>';
    die();
}else{
    echo'
    <div class="panel panel-success">
        <div class="panel-heading">Respuesta Transacción</div>
            <div class="panel-body">
                <div class="row">
                    <div class="col-xs-7 col-md-7">
                        <h3>Respuesta del Banco</h3>
                        <p>
                            <strong>Mensaje del Banco</strong>: '.$msg_banco.'<br/>
                            <strong>ID del Pago</strong>: '.$id_pago.'<br/>
                            <strong>ID Referencia</strong>: '.$reference.'<br/>
                        </p>
                    </div>
                    <div class="col-xs-5 col-md-5 pull-right">
                        <h3>Voucher</h3>
                        '.$voucher.'
                    </div>
                </div>
            </div>
        </div>
    </div>';
}

?>
